src/Protocols/Axi4/Strict/Full/ReadAddress.hs

```diff
@@ -0,0 +1,186 @@
+{-|
+Defines ReadAddress channel of full AXI4 protocol with port names corresponding
+to the AXI4 specification.
+-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -Wno-missing-fields #-}
+
+module Protocols.Axi4.Strict.Full.ReadAddress
+  ( M2S_ReadAddress(..)
+  , S2M_ReadAddress(..)
+  , Axi4ReadAddress
+  ) where
+
+-- base
+import Data.Coerce
+import Data.Kind (Type)
+import Data.Proxy
+import GHC.Generics (Generic)
+
+-- clash-prelude
+import qualified Clash.Prelude as C
+
+-- me
+import Protocols.Axi4.Common
+import Protocols.Internal
+import Protocols.DfLike (DfLike)
+import qualified Protocols.DfLike as DfLike
+
+-- | AXI4 Read Address channel protocol
+data Axi4ReadAddress
+  (dom :: C.Domain)
+  (kb :: KeepBurst)
+  (ksz :: KeepSize)
+  (lw :: LengthWidth)
+  (iw :: IdWidth)
+  (aw :: AddrWidth)
+  (kr :: KeepRegion)
+  (kbl :: KeepBurstLength)
+  (kl :: KeepLock)
+  (kc :: KeepCache)
+  (kp :: KeepPermissions)
+  (kq :: KeepQos)
+  (userType :: Type)
+
+instance Protocol (Axi4ReadAddress dom kb ksz lw iw aw kr kbl kl kc kp kq userType) where
+  type Fwd (Axi4ReadAddress dom kb ksz lw iw aw kr kbl kl kc kp kq userType) =
+    C.Signal dom (M2S_ReadAddress kb ksz lw iw aw kr kbl kl kc kp kq userType)
+  type Bwd (Axi4ReadAddress dom kb ksz lw iw aw kr kbl kl kc kp kq userType) =
+    C.Signal dom S2M_ReadAddress
+
+instance Backpressure (Axi4ReadAddress dom kb ksz lw iw aw kr kbl kl kc kp kq userType) where
+  boolsToBwd _ = C.fromList_lazy . coerce
+
+instance DfLike dom (Axi4ReadAddress dom kb ksz lw iw aw kr kbl kl kc kp kq) userType where
+  type Data (Axi4ReadAddress dom kb ksz lw iw aw kr kbl kl kc kp kq) userType =
+    M2S_ReadAddress kb ksz lw iw aw kr kbl kl kc kp kq userType
+
+  type Payload userType = userType
```

Comment:
After working some more with AXI, I don't think exposing `userType` as the payload is a good idea. Perhaps we should define the address as the payload, so functions like `map` operate on that instead.